File: puzzles/haskell/defibrillators.hs
import System.IO
import Control.Monad
import Text.Regex

main :: IO ()
main = do
  hSetBuffering stdout NoBuffering -- DO NOT REMOVE
  
  input_line <- getLine
  let lon = toRadians(input_line :: String)
  input_line <- getLine
  let lat = toRadians(input_line :: String)
  input_line <- getLine
  let n = read input_line :: Int
  
  defibrillators <- replicateM n $ do
    d <- getLine
    let defib = splitRegex (mkRegex ";") d
    let name = defib !! 1
    let defibLon = toRadians(defib !! 4)
    let defibLat = toRadians(defib !! 5)
    let x = (defibLon - lon) * cos((lat + defibLat) / 2)
    let y = defibLat - lat
    let distance = sqrt(x^2 + y^2) * 6371
    return (distance, name)
  
  let closestDefib = minimum defibrillators
  putStrLn (snd closestDefib)
  return ()

-- Convert degrees to radians
toRadians :: String -> Double
toRadians angle = (read (map replace angle) :: Double) * pi / 180

replace :: Char -> Char
replace ',' = '.'
replace c = c

File: puzzles/python3/defibrillators.py
import math


def to_radians(distance):
    """ Convert distance in degrees to radians. """
    return float(distance.replace(",", ".", 1)) * math.pi / 180


min_dist = float("inf")
name = ""
lon = to_radians(input())
lat = to_radians(input())
n = int(input())

# Determine the name of the defib closest to the user’s position.
for i in range(n):
    defib = input().split(";")
    defib_lon = to_radians(defib[4])
    defib_lat = to_radians(defib[5])
    x = (defib_lon - lon) * math.cos((lat + defib_lat) / 2)
    y = defib_lat - lat
    distance = math.hypot(x, y) * 6371
    if distance < min_dist:
        min_dist = distance
        name = defib[1]

print(name)
